src/magic_folder/snapshot.py

```diff
@@ -0,0 +1,172 @@
+# Copyright 2020 Least Authority TFA GmbH
+# See COPYING for details.
+
+"""
+Functions and types that implement snapshots
+"""
+from __future__ import print_function
+
+import time
+import json
+
+from itertools import count
+
+from twisted.internet.defer import (
+    inlineCallbacks,
+    returnValue,
+)
+
+from .common import (
+    get_node_url,
+)
+
+from twisted.web.client import (
+    Agent,
+    readBody,
+)
+
+from twisted.web.http import (
+    OK,
+    CREATED,
+)
+
+from treq.client import (
+    HTTPClient,
+)
+
+from hyperlink import (
+    DecodedURL,
+)
+
+from .common import (
+    bad_response,
+)
+
+
+# version of the snapshot scheme
+SNAPSHOT_VERSION = 1
+
+@inlineCallbacks
+def tahoe_put_immutable(nodeurl, filepath, treq):
+    """
+    :param unicode nodeurl: The web endpoint of the Tahoe-LAFS client
+        associated with the magic-folder client.
+    :param unicode filepath: The file path that needs to be stored into
+        the grid.
+    :param HTTPClient treq: An ``HTTPCLient`` or similar object to use to
+        make the queries.
+    :return Deferred[unicode]: The readcap associated with the newly created
+        unlinked file.
+    """
+    node_url = DecodedURL.from_text(unicode(nodeurl, 'utf-8'))
+    url = node_url.child(
+        u"uri",
+    ).add(
+        u"format",
+        u"CHK",
+    )
+
+    # XXX: check whether we need to set any headers.
+    put_uri = url.to_uri().to_text().encode("ascii")
+    with open(str(filepath), "rb") as file:
+        data = file.read()
+
+    response = yield treq.put(put_uri, data)
+    if response.code == OK or response.code == CREATED:
+        result = yield readBody(response)
+        returnValue(result)
+    else:
+        raise Exception("Error response PUT {} - {}".format(put_uri, response))
+
+@inlineCallbacks
+def tahoe_create_snapshot_dir(nodeurl, content, parents, timestamp, treq):
+    """
+    :param unicode content: readcap for the content.
+    :param [unicode] parents: List of parent snapshot caps
+    :param unicode author: readcap that represents the author pubkey
+    :param integer timestamp: POSIX timestamp that represents the creation time
+    :return Deferred[unicode]: The readcap associated with the newly created
+        snapshot.
+    """
```

Comment:
It would probably be useful to have a structured representation of most/all of this information as a Python object so that there is a thing that can be pointed at and called a "snapshot" rather than needing to manage all of a snapshots fields as separate parameters/variables.